Added rack_elevation.show_images to user.preferences

ogorov opened the pull request:
### Fixes: [5203](https://github.com/netbox-community/netbox/issues/5203)

- Changed rack_elevation.js to show/hide images in case of preference is set and to send preference to server when user toggle it
- Added post method to RackElevationListView (I can't find better place to add method that set new preference)
- Made toggle-button glyph and attributes dependent on new property

Commit:
Added rack_elevation.show_images to user.preferences

Changed rack_elevation.js to show/hide images in case of prefernce is set and to send preference to server when user toggle it
Added post method to RackElevationListView.
(I can't find better place to add method that set new preference)
Made toggle-button gliph and attributes dependent on new property

ogorov commented:
I don't sure if it matches the style of NetBox. But I'm ready to make changes if in will needed.

jeremystretch commented on netbox/dcim/views.py:
```diff
@@ -341,8 +342,21 @@ def get(self, request):
             'reverse': reverse,
             'rack_face': rack_face,
             'filter_form': forms.RackElevationFilterForm(request.GET),
+            'show_images': request.user.config.get('rack_elevation.show_images', False)
```
Images should be enabled by default to preserve the existing behavior.

ogorov replied:
It will be easy to change

jeremystretch commented on netbox/dcim/views.py:
```diff
@@ -341,8 +342,21 @@ def get(self, request):
+            'show_images': request.user.config.get('rack_elevation.show_images', False)
         })
 
+    def post(self, request):
```
Any such dynamic updates would need to use the REST API. (Adding individual view methods for all of the various settings does not scale.) Unfortunately, we don't yet have a REST API endpoint for user preferences. I believe this work is blocked until one can be introduced.

ogorov replied:
Has a suitable problem been created or do I need to create a new one?

jeremystretch replied:
The upcoming v2.10 release introduces an endpoint for this. If you want, you can rebase this PR off the `develop-2.10` branch and adapt to use the new endpoint, or just wait for the release (which should be sometime next week).

ogorov replied:
OK, I'll do something about it soon.

DanSheps replied:
Do you have an update for this?

Commit:
Merge branch 'develop' into develop-2.10

jeremystretch commented:
With v2.10 released, the `/api/users/config/` endpoint is now available for use.

ogorov commented:
> With v2.10 released, the `/api/users/config/` endpoint is now available for use.

I tried to set preference with new api endpoint. But I can't set values as a hierarchical dictionary structure like other preferences. This is because PATCH method uses the deepmerge function, which adds the raw received params to the user preferences. The raw data can only be a string, but I want to set dictionary. 
There is also no protection from passing bad data (non-existent field) to the user preferences. 
Is changing the logic of the PATCH method a good idea or am I missing something?

Commit:
Fixes with new API endpoint
Rewrote js to send data to the new endpoint
Made preferences API endpoint accept JSON structures

Commit:
Merge remote-tracking branch 'True-NetBox/develop' into develop-2.10

Commit:
Add exception type for try catch of api method

Commit:
Fix exceptions in API

ogorov committed:
Fixes for anonimous user
Added check if user is authenticated
Deleted debbuging output

ogorov committed:
Fix

ogorov commented:
Are there any comments about this pull request?

jeremystretch commented:
It seems like this has gotten pretty far off track. Maybe it makes sense to defer this work until the user preferences API endpoint becomes a bit more mature. We shouldn't be trying to address the limitations you cite above as part of this FR.

ogorov commented:
Hi
Is there any sense on this issue?

jeremystretch commented:
Looks like this should have been closed out a while back. As noted above, we need more work around the preferences architecture before we consider extending it any further. Additionally, the UI has been completed reworked for v3.0.
